### ISSUE TITLE
[RFC]: Support KV Cache CPU Offloading

### ISSUE THREAD

**Issue by Vivo50E** [vllm-project/vllm-omni#1150]
### Motivation.

vLLM‑Omni pipelines often combine autoregressive (AR) LLM stages with vision/audio encoders and diffusion decoders, where AR KV cache quickly dominates GPU memory for long‑context and concurrent workloads.  
Today, AR KV must remain fully on GPU, forcing operators to shrink `max_model_len` / `max_num_seqs`, scale up GPU capacity, or tolerate OOM‑driven failures.  

Omni already offloads diffusion weights to CPU, but this mechanism does not apply to AR KV, and there is currently no AR KV→CPU path in Omni.  

vLLM v1 is introducing an OffloadingConnector/OffloadingManager for async, block‑aligned KV offload to CPU and other media, but it is still evolving and not integrated with Omni’s multi‑stage scheduler.

This RFC proposes an Omni‑local mechanism to offload AR KV blocks from GPU to CPU and prefetch them back when needed, integrate offload decisions with Omni’s scheduler and AR stage runner, reuse LMCache as the initial CPU KV backend, and preserve a clean migration path toward future alignment with vLLM v1’s OffloadingConnector.

## Target

**Features:**
- Enable KV cache CPU offloading for AR LLM stages in vLLM‑Omni.
- Trade latency for memory: longer context lengths and/or higher concurrency on fixed HBM.
- Keep most logic in Omni; minimal vLLM core changes.
- Simple user configuration (on/off + key knobs) with forward compatibility to vLLM v1's OffloadingConnector.
- Metrics and debug tooling for KV utilization and offload/prefetch behavior.

**In Scope:**
- AR LLM stages (single‑GPU/worker MVP); integration into `GPUARModelRunner` and scheduler.
- Per‑worker GPU↔CPU KV tiering (block‑level), with Omni deciding *when* to offload and LMCache handling *how*.

**Out of Scope:**
- Diffusion / DiT stages and their internal block caches.
- Multi‑GPU / multi‑node tiered KV beyond what Omni already does for AR.
- vLLM v1 native CPU offloading backend (planned as follow‑up).

### Accuracy

* Correctness: Offloaded KV blocks must exactly match original GPU KV (no numerical corruption or truncation).
* Behavior: Same prompts and settings must yield identical tokens with and without offload (modulo expected nondeterminism like sampling).
* Edge cases: Hitting GPU threshold, max CPU blocks, context length limits, or rapid on/off toggling must not change model outputs, only performance.
* Compatibility: Must work with existing scheduler/pipeline configs and KV layout APIs without requiring model or checkpoint changes.
* Metrics: Validate with golden prompts comparing logits/token sequences before/after offload, and add regression tests for long‑context and OOM‑near scenarios.

### Performance

* Do no harm when offload is effectively inactive
* Provide a controlled, tunable slowdown when trading latency for memory,
* Avoid hangs or unbounded CPU RAM growth like those reported in earlier CPU offload issues, especially in complex Omni pipelines mixing AR and diffusion.

### Concurrency & thread safety

OmniCpuKvStore must be safe under multi-request concurrent access, with a clearly defined concurrency model (e.g., multi-reader/single-writer or fully concurrent) and internal synchronization (global lock, sharded locks, or lock-free data structures) that does not become a bottleneck under typical vLLM workloads.

### High-level Design

The proposed design adds an Omni‑specific KV offload layer around AR stages with three main pieces:

1. **Configuration**  

   - New fields on Omni engine / stage configs to enable and tune KV offload.

2. **Scheduler‑side offload manager (`OmniKvOffloadManager`)**  

   - Inspects AR KV usage per scheduling step.  

   - Generates a `SwapPlan` describing GPU→CPU offloads and CPU→GPU prefetches for specific requests and KV blocks.  

   - Attaches this `SwapPlan` to `SchedulerOutput`.

3. **Worker‑side KV store (`OmniCpuKvStore`) and runner integration**  

   - `GPUARModelRunner` executes the `SwapPlan` at the beginning of `execute_model`, using `OmniCpuKvStore` as the CPU KV backend.  

   - MVP backend is a LMCache‑based CPU KV store. 
   
   - A vLLM v1 native offloading backend can be added later under the same  OmniCpuKvStore  abstraction.

  
####  Components

* OmniKvOffloadManager  (scheduler side)
* SwapPlan  (metadata on  SchedulerOutput )	
* GPUARModelRunner  (worker side)
* OmniCpuKvStore  interface +  LmcacheKvStore  implementation
* Existing vLLM KV cache manager / tensors

#### Components Interaction

```mermaid
%%{init: {'theme':'default','flowchart':{'diagramPadding':5,'htmlLabels':true},'themeVariables':{'fontSize':'10px'}}}%%
flowchart TB
    C[Client] --> S[OmniScheduler]

    subgraph scheduler[Scheduler Side]
        S --> OM[OffloadMgr]
        OM --> S
    end

    S --> SO[[SchedulerOutput + SwapPlan]]

    subgraph worker[Worker Side]
        SO --> R[ARRunner]
        R --> KTM[kv_transfer]
    end

    subgraph gpu[GPU]
        K[KV Cache]
        W[AR Weights]
    end

    subgraph cpu[CPU]
        P[CpuKvStore]
    end

    R -->|put/get| P
    R -->|KV rw| K
    P <--> K

```

###  Configuration & Engine Wiring

We extend Omni’s engine and stage configuration so that AR stages can opt‑in to KV CPU offloading and choose a backend (MVP: LMCache).

```python
@dataclass
class OmniEngineArgs(EngineArgs):
    # existing fields ...

    enable_cpu_kv_offload: bool = False
    cpu_kv_offload_config: Optional[Dict[str, Any]] = None
    # Example:
    # {
    #   "backend": "lmcache",        # MVP: only "lmcache" is supported
    #   "gpu_usage_threshold": 0.8,  # start offloading when GPU KV > 80%
    #   "max_cpu_blocks": 200000,    # soft upper bound on CPU KV blocks
    # }
```

- When `enable_cpu_kv_offload=false` (default), Omni behavior is unchanged.  
- When `enable_cpu_kv_offload=true`, the AR stage:
  - constructs an `OmniKvOffloadManager` on the scheduler side, and  
  - creates an `OmniCpuKvStore` implementation (LMCache) on the worker side.

The config is passed through:

1. CLI / YAML → `OmniEngineArgs`.  
2. `OmniEngineArgs` → AR stage config (`OmniARStageConfig`, or equivalent).  
3. AR stage config → scheduler & worker constructors.

For LMCache, we also rely on the standard LMCache configuration mechanism (e.g. `LMCACHE_CONFIG_FILE`, chunk size, CPU RAM budget, etc.) as described in LMCache docs.

### OmniKvOffloadManager (Scheduler Side)

`OmniKvOffloadManager` is responsible for translating **global KV usage** into a per‑step `SwapPlan`.

#### Inputs

On each scheduler iteration, the manager consumes:

- **Per‑request KV metadata**, including at least:
	- approximate number of KV blocks (or bytes) currently on GPU per `req_id`,  
	- a notion of “age” (e.g., earliest token index or last‑access step).  
- **Global GPU KV usage**:
	- either as bytes, or as fraction of a configured max KV budget.  
- **Configuration**:
	- `gpu_usage_threshold`  
	- `max_cpu_blocks`  

The KV metadata can either come from:

- vLLM’s KV manager and `Request` structures (e.g., block IDs and `block_hashes`), or  
- an Omni‑side shadow structure maintained alongside scheduler state.

We do **not** change core vLLM KV layout; we just read existing metadata.

#### Output: SwapPlan

The manager returns a `SwapPlan` for the next step:

```python
class SwapPlan(NamedTuple):
    gpu_to_cpu: Dict[ReqId, List[BlockId]]
    cpu_to_gpu: Dict[ReqId, List[BlockId]]
```

- `gpu_to_cpu`:
  - list of blocks to be evicted from GPU to CPU before this step’s forward pass.  
- `cpu_to_gpu`:
  - list of blocks that must be prefetched from CPU to GPU so that the AR model forward can read them during this step.

The `SwapPlan` is attached to `SchedulerOutput` for the AR worker. If the plan is empty, the worker simply skips offload/preload.

#### Policy (MVP)

The MVP keeps vLLM’s offload policy minimal and delegates detailed eviction decisions to LMCache:

1. vLLM only decides **when** to offload/prefetch based on `gpu_kv_usage` and `gpu_usage_threshold`, and constructs a `SwapPlan` containing candidate blocks per request.  
2. LMCache receives this `SwapPlan` and applies its own backend policy (e.g., per‑request sliding window, LRU, TTL, multi‑tier placement) to choose which KV blocks to keep or evict on each tier (GPU / CPU / local storage).  
3. Prefetch (`cpu_to_gpu`) is likewise driven by LMCache’s policy (e.g., next‑needed blocks for active decoding), while vLLM only ensures that the requested blocks are available on GPU before attention is computed.

### OmniCpuKvStore & LMCache Integration (Worker Side)

`OmniCpuKvStore` defines the worker‑side abstraction for CPU KV storage and transfer. Its core responsibility is to map `(req_id, block_ids)` onto the underlying CPU tier (LMCache in MVP).

#### Interface

```python
class OmniCpuKvStore(Protocol):
    def put(
        self,
        req_id: ReqId,
        block_ids: Sequence[BlockId],
        kv_view: "KvView",
    ) -> None:
        """
        Offload specified KV blocks from GPU to CPU tier.
        kv_view abstracts how to locate GPU KV tensors and indices.
        """

    def get(
        self,
        req_id: ReqId,
        block_ids: Sequence[BlockId],
        kv_view: "KvView",
    ) -> None:
        """
        Prefetch specified KV blocks from CPU tier back to GPU.
        """
```

`KvView` is a small helper that encapsulates access to the vLLM KV cache tensors for a given `req_id`:

- knows how to map a `BlockId` to a slice in the underlying `kv_caches` tensor,  
- hides attention backend details (FlashAttention / FlashInfer, etc.).

#### LmcacheKvStore (MVP Backend)

`LmcacheKvStore` implements `OmniCpuKvStore` using LMCache:

- It holds a LMCache client / connector (e.g., configured via `kv_transfer_config` with `kv_connector="LMCacheConnectorV1"`).
- `put`:
  - For each `block_id`:
    - uses `kv_view` to extract the corresponding KV slice from GPU tensor;  
    - calls LMCache APIs (e.g., `save_kv` / `put_block`) to store the KV block in the CPU RAM backend (as configured by LMCache).  
- `get`:
  - For each `block_id`:
    - calls LMCache APIs (e.g., `load_kv` / `get_block`) to retrieve the block;  
    - writes it back into the appropriate region of the GPU KV tensor via `kv_view`.

#### Backend Factory

We introduce a simple factory to construct `OmniCpuKvStore`:

```python
class OmniConnectorFactory:
    @classmethod
    def create_cpu_kv_store(cls, cfg: Dict[str, Any]) -> OmniCpuKvStore:
        backend = cfg.get("backend", "lmcache")
        if backend == "lmcache":
            return LmcacheKvStore(cfg)
        else:
            raise ValueError(f"Unsupported cpu_kv_offload backend: {backend}")
```

Initially, only `"lmcache"` is supported. A future extension can add `"v1_native"` which wraps vLLM v1 OffloadingConnector’s CPU backend.

***

### GPUARModelRunner Integration

`GPUARModelRunner` is the worker‑side entry point for AR stage execution. We extend it to:

1. Optionally hold an `OmniCpuKvStore`.  
2. Execute the `SwapPlan` before the usual KV transfer and forward pass.

#### Initialization

```python
class GPUARModelRunner(OmniGPUModelRunner):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.cpu_kv_store: Optional[OmniCpuKvStore] = None
        if getattr(self.model_config, "enable_cpu_kv_offload", False):
            cfg = getattr(self.model_config, "cpu_kv_offload_config", {}) or {}
            self.cpu_kv_store = OmniConnectorFactory.create_cpu_kv_store(cfg)
```

#### SwapPlan Execution

```python
    @torch.inference_mode()
    def execute_model(self, scheduler_output, intermediate_tensors=None):
        # 1. Execute SwapPlan (if any) before existing logic
        swap_plan = getattr(scheduler_output, "swap_plan", None)
        if swap_plan and self.cpu_kv_store is not None:
            self._execute_swap_plan(swap_plan)

        # 2. Existing cross‑stage KV transfer
        self.kv_extracted_req_ids = (
            self.kv_transfer_manager.handle_finished_requests_kv_transfer(...)
        )

        # 3. Existing model forward pass (unchanged)
        # ...

    def _execute_swap_plan(self, swap_plan: SwapPlan) -> None:
        # GPU → CPU offload
        for req_id, block_ids in swap_plan.gpu_to_cpu.items():
            kv_view = self._kv_view_for(req_id)
            self.cpu_kv_store.put(req_id, block_ids, kv_view)

        # CPU → GPU prefetch
        for req_id, block_ids in swap_plan.cpu_to_gpu.items():
            kv_view = self._kv_view_for(req_id)
            self.cpu_kv_store.get(req_id, block_ids, kv_view)
```

Key points:

- `_kv_view_for(req_id)` encapsulates how to reach the underlying vLLM KV tensors for that request.  
- We run `_execute_swap_plan` **before** the existing inter‑stage KV transfer and model forward, so that:
  - all needed KV is present on GPU for this step’s attention,  
  - evicted KV has already been safely stored in LMCache.

#### Interaction with Existing KV Transfer

`kv_transfer_manager.handle_finished_requests_kv_transfer(...)` continues to manage KV movement across stages/workers (e.g., disaggregated prefill). KV offloading is orthogonal:

- Offload decisions must avoid touching KV that is still needed by other stages in this step.  
- In the MVP, we can enforce a simple rule: only offload KV for requests that are “owned” exclusively by this AR stage at scheduling time.

Longer‑term, we can consider tighter integration so that `OmniKvOffloadManager` is aware of cross‑stage dependencies.

***

### Data & Control Flow Summary

Putting it together for one scheduler step:

1. **Scheduler side**  
   - `OmniScheduler` collects per‑request KV stats and global GPU KV usage.  
   - Calls `OmniKvOffloadManager.update_and_build_swap_plan(...)`.  
   - Receives a `SwapPlan` and attaches it to `SchedulerOutput` for the AR worker.

2. **Worker side – pre‑forward**  
   - `GPUARModelRunner.execute_model` reads `swap_plan`.  
   - Calls `_execute_swap_plan`:
     - For each `(req_id, block_ids)` in `gpu_to_cpu`:  
       - `LmcacheKvStore.put` extracts KV slices and passes them to LMCache (GPU→CPU).  
     - For each `(req_id, block_ids)` in `cpu_to_gpu`:  
       - `LmcacheKvStore.get` loads KV from LMCache and writes them into GPU KV tensors (CPU→GPU).

3. **Worker side – existing logic**  
   - Runs inter‑stage KV transfer as before.  
   - Runs AR forward pass, where attention kernels **directly read current GPU KV tensors**; KV that remains offloaded to CPU is invisible until a future `SwapPlan` triggers prefetch.

4. **State updates**  
   - After forward, KV metadata (e.g., new blocks created) is updated in scheduler state for the next iteration’s offload decisions.

### Use case 1: Long‑context chat on small GPUs

**Problem:** Long conversations on small GPUs (12–24 GB HBM) hit OOM, forcing conservative `max_model_len`/`max_num_seqs`.  
**Solution:** Enable KV offloading (`gpu_usage_threshold: 0.8`). Oldest blocks offload to CPU; recent/active KV stays on GPU.  
**Impact:** Longer context, higher concurrency, moderate latency trade-off.

### Use case 2: Multi‑stage omni pipeline colocation

**Problem:** KV cache for AR LLM competes with diffusion weights/activations on shared GPU.  
**Solution:** AR stage offloads old KV blocks to CPU when diffusion is active.  
**Impact:** Better HBM sharing; larger diffusion models or more steps without additional GPUs.

### Use case 3: Offline batch processing of long documents

**Problem:** Long documents require large GPUs or aggressive chunking (losing context).  
**Solution:** Configure aggressive offload policy (low threshold, large `max_cpu_blocks`).  
**Impact:** Process much longer documents in single pass; latency acceptable for batch jobs.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by princepride** [comment]
Well done. It appears that there will be no conflict with the KV cache allocated for stage-to-stage transfers. I will conduct a more thorough review later.

**Comment by SamitHuang** [comment]
LGTM overall

**Comment by hsliuustc0106** [comment]
@Vivo50E are you willing to contribute on this one? maybe you can share your thoughts in slack first.

**Comment by Vivo50E** [comment]
> [@Vivo50E](https://github.com/Vivo50E) are you willing to contribute on this one? maybe you can share your thoughts in slack first.

@hsliuustc0106 Yes, I'm interested in trying this out. I’m happy to attend weekly meetings and come up with an initial implementation PR.

**Comment by KuntaiDu** [comment]
Hi! I am a committer from vLLM & LMCache side, happy to chat more and collaborate!
cc @ApostaC @sammshen @allytotheson

**Comment by Vivo50E** [comment]
Hello Kuntai @KuntaiDu , I’d be glad to discuss this further and collaborate with you. I’m not an expert in LMCache or KV-cache optimization, but I’d love to hear your thoughts on this RFC. I’m currently working on an initial PR to implement the feature—any feedback or guidance would be greatly appreciated.

**Comment by hsliuustc0106** [comment]
> Hi! I am a committer from vLLM & LMCache side, happy to chat more and collaborate! cc [@ApostaC](https://github.com/ApostaC) [@sammshen](https://github.com/sammshen) [@allytotheson](https://github.com/allytotheson)

@KuntaiDu please check #1330 for an initial trial from @Vivo50E
